File: sdk-typescript/packages/zoom/src/zoom.ts
/**
 * WAVE SDK - Zoom API
 *
 * Zoom meeting, Zoom Room, and RTMS integration.
 */

import type { WaveClient, PaginationParams, PaginatedResponse, Timestamps } from '@wave-av/core';

export interface ZoomMeeting {
  id: string;
  topic: string;
  type: "instant" | "scheduled" | "recurring";
  status: "waiting" | "started" | "ended";
  start_url: string;
  join_url: string;
  host_id: string;
  duration_minutes: number;
  participants_count: number;
  recording_enabled: boolean;
  rtms_enabled: boolean;
  created_at: string;
}
export interface ZoomRoom {
  id: string;
  name: string;
  location?: string;
  status: "online" | "offline" | "in_meeting";
  account_id: string;
  device_ip?: string;
  firmware_version?: string;
  camera_count: number;
  microphone_count: number;
}
export interface ZoomRecording extends Timestamps {
  id: string;
  meeting_id: string;
  type: "cloud" | "rtms";
  status: "processing" | "completed" | "failed";
  file_url?: string;
  duration_seconds: number;
  file_size_bytes: number;
}
export interface CreateMeetingRequest {
  topic: string;
  type?: "instant" | "scheduled" | "recurring";
  duration_minutes?: number;
  start_time?: string;
  recording_enabled?: boolean;
  rtms_enabled?: boolean;
  password?: string;
}
export interface ListMeetingsParams extends PaginationParams {
  status?: string;
  type?: string;
  host_id?: string;
}

/**
 * Zoom meeting, Zoom Room, and RTMS streaming integration.
 *
 * @example
 * ```typescript
 * const meeting = await wave.zoom.createMeeting({ topic: "Team Standup" });
 * await wave.zoom.startRTMS(meeting.id, { stream_url: "rtmp://...", stream_key: "key" });
 * const rooms = await wave.zoom.listRooms();
 * ```
 */
export class ZoomAPI {
  private readonly client: WaveClient;
  private readonly basePath = "/v1/zoom";
  constructor(client: WaveClient) {
    this.client = client;
  }

  async createMeeting(request: CreateMeetingRequest): Promise<ZoomMeeting> {
    return this.client.post<ZoomMeeting>(`${this.basePath}/meetings`, request);
  }
  async getMeeting(meetingId: string): Promise<ZoomMeeting> {
    return this.client.get<ZoomMeeting>(`${this.basePath}/meetings/${meetingId}`);
  }
  async endMeeting(meetingId: string): Promise<void> {
    await this.client.post(`${this.basePath}/meetings/${meetingId}/end`);
  }
  async listMeetings(params?: ListMeetingsParams): Promise<PaginatedResponse<ZoomMeeting>> {
    return this.client.get<PaginatedResponse<ZoomMeeting>>(`${this.basePath}/meetings`, {
      params: params as Record<string, string | number | boolean | undefined>,
    });
  }
  async listRooms(params?: PaginationParams): Promise<PaginatedResponse<ZoomRoom>> {
    return this.client.get<PaginatedResponse<ZoomRoom>>(`${this.basePath}/rooms`, {
      params: params as Record<string, string | number | boolean | undefined>,
    });
  }
  async getRoomStatus(roomId: string): Promise<ZoomRoom> {
    return this.client.get<ZoomRoom>(`${this.basePath}/rooms/${roomId}`);
  }
  async getRecording(recordingId: string): Promise<ZoomRecording> {
    return this.client.get<ZoomRecording>(`${this.basePath}/recordings/${recordingId}`);
  }
  async listRecordings(
    meetingId?: string,
    params?: PaginationParams,
  ): Promise<PaginatedResponse<ZoomRecording>> {
    const path = meetingId
      ? `${this.basePath}/meetings/${meetingId}/recordings`
      : `${this.basePath}/recordings`;
    return this.client.get<PaginatedResponse<ZoomRecording>>(path, {
      params: params as Record<string, string | number | boolean | undefined>,
    });
  }
  async startRTMS(
    meetingId: string,
    config: { stream_url: string; stream_key: string },
  ): Promise<{ status: string }> {
    return this.client.post<{ status: string }>(
      `${this.basePath}/meetings/${meetingId}/rtms/start`,
      config,
    );
  }
  async stopRTMS(meetingId: string): Promise<{ status: string }> {
    return this.client.post<{ status: string }>(`${this.basePath}/meetings/${meetingId}/rtms/stop`);
  }
  async getRTMSStatus(meetingId: string): Promise<{ status: string; stream_url?: string }> {
    return this.client.get<{ status: string; stream_url?: string }>(
      `${this.basePath}/meetings/${meetingId}/rtms`,
    );
  }
}

export function createZoomAPI(client: WaveClient): ZoomAPI {
  return new ZoomAPI(client);
}

File: sdk-python/wave_av_sdk/zoom.py
"""WAVE SDK - Zoom API. Zoom meetings, rooms, recordings, and RTMS."""
from __future__ import annotations
from typing import Any
from pydantic import BaseModel
from wave.client import WaveClient

class ZoomMeeting(BaseModel):
    id: str; topic: str; type: str; status: str; start_url: str; join_url: str; host_id: str; duration_minutes: int; participants_count: int = 0; recording_enabled: bool = False; rtms_enabled: bool = False; created_at: str

class ZoomRoom(BaseModel):
    id: str; name: str; location: str | None = None; status: str; account_id: str; camera_count: int = 0; microphone_count: int = 0

class ZoomRecording(BaseModel):
    id: str; meeting_id: str; type: str; status: str; file_url: str | None = None; duration_seconds: int = 0; file_size_bytes: int = 0; created_at: str

class ZoomAPI:
    def __init__(self, client: WaveClient): self._client = client; self._base = "/v1/zoom"
    def create_meeting(self, topic: str, **kwargs: Any) -> ZoomMeeting: return ZoomMeeting(**self._client.post(f"{self._base}/meetings", json={"topic": topic, **kwargs}))
    def get_meeting(self, meeting_id: str) -> ZoomMeeting: return ZoomMeeting(**self._client.get(f"{self._base}/meetings/{meeting_id}"))
    def end_meeting(self, meeting_id: str) -> None: self._client.post(f"{self._base}/meetings/{meeting_id}/end")
    def list_meetings(self, **params: Any) -> dict: return self._client.get(f"{self._base}/meetings", params={k: v for k, v in params.items() if v is not None})
    def list_rooms(self, **params: Any) -> dict: return self._client.get(f"{self._base}/rooms", params={k: v for k, v in params.items() if v is not None})
    def get_room_status(self, room_id: str) -> ZoomRoom: return ZoomRoom(**self._client.get(f"{self._base}/rooms/{room_id}"))
    def get_recording(self, recording_id: str) -> ZoomRecording: return ZoomRecording(**self._client.get(f"{self._base}/recordings/{recording_id}"))
    def list_recordings(self, meeting_id: str | None = None, **params: Any) -> dict: path = f"{self._base}/meetings/{meeting_id}/recordings" if meeting_id else f"{self._base}/recordings"; return self._client.get(path, params={k: v for k, v in params.items() if v is not None})
    def start_rtms(self, meeting_id: str, stream_url: str, stream_key: str) -> dict: return self._client.post(f"{self._base}/meetings/{meeting_id}/rtms/start", json={"stream_url": stream_url, "stream_key": stream_key})
    def stop_rtms(self, meeting_id: str) -> dict: return self._client.post(f"{self._base}/meetings/{meeting_id}/rtms/stop")
    def get_rtms_status(self, meeting_id: str) -> dict: return self._client.get(f"{self._base}/meetings/{meeting_id}/rtms")
